### PR TITLE
Bracketed escaping

### PR DESCRIPTION
**Pull request by jayvdb**
Adds `{\}`, `{{}` and `{}}` as an escaping mechanism in settings.

`{{}` and `{}}` are not implemented so they can be used safely together.
Additional coding is needed to handle values from the engine being pushed
back into the engine, so they are re-escaped.

Fixes https://github.com/tox-dev/tox/issues/1708

## Contribution checklist:

(also see [CONTRIBUTING.rst](../tree/master/CONTRIBUTING.rst) for details)

- [ ] wrote descriptive pull request text
- [x] added/updated test(s)
- [ ] updated/extended the documentation
- [ ] added relevant [issue keyword](https://help.github.com/articles/closing-issues-using-keywords/)
      in message body
- [ ] added news fragment in [changelog folder](../tree/master/docs/changelog)
  * fragment name: `<issue number>.<type>.rst` for example (588.bugfix.rst)
  * `<type>` is must be one of `bugfix`, `feature`, `deprecation`,`breaking`, `doc`, `misc`
  * if PR has no issue: consider creating one first or change it to the PR number after creating the PR
  * "sign" fragment with "by :user:`<your username>`"
  * please use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files - by :user:`superuser`."
  * also see [examples](../tree/master/docs/changelog)
- [x] added yourself to `CONTRIBUTORS` (preserving alphabetical order)

### CODE REVIEW

File: tests/unit/config/test_config.py [hunk 1]
```diff
@@ -2803,6 +2803,47 @@ def test_multiple_homedir_relative_local_indexservers(self, newconfig):
...
+        config = newconfig(
+            """
+        [testenv]
+        setenv =
+            FOO = a{{}b c{}}d
+        """,
+        )
+        envconfig = config.envconfigs["python"]
+        # This fails: ConfigError: substitution key 'b c' not found
+        # assert envconfig.setenv["FOO"] == "a{b c}d"
```
Comment (jayvdb) [inline tests/unit/config/test_config.py:2845]: highlighting this as what fails with the incomplete implementation.

**Comment by gaborbernat** [comment]
> Adds `{\}`, `{{}` and `{}}` as an escaping mechanism in settings.
> 
> `{{}` and `{}}` are not implemented so they can be used safely together.
> Additional coding is needed to handle values from the engine being pushed
> back into the engine, so they are re-escape

I would prefer not to introduce new ways of doing things. As such only ``\`` character I'm interested in supporting, I don't agree with supporting ``{{`` or ``}}`` 🤔

**Comment by jayvdb** [comment]
> I don't agree with supporting `{{` or `}}`

I am also not convinced about adding those; I've tried to lay out the problems at

https://github.com/tox-dev/tox/issues/1708

That commit in this PR is mostly to show where the problem exists, so anyone can play with it if they think it is easy to solve.

**Review by gaborbernat — CHANGES_REQUESTED** [review]
This now has merge conflicts.

**Comment by gaborbernat** [comment]
@jayvdb this is merge conflicting.

**Comment by gaborbernat** [comment]
I'll close this for now until you can attend to the PR again. Once you've done we can reopen.

**Pull request closed without merging**